curriculum/challenges/english/01-responsive-web-design/basic-html-and-html5/inpaint.md

```diff
@@ -0,0 +1,213 @@
+<section id='description'>
+"Inpaint" method will help you restore your damaged picture via using one of the two algorithms: Fast Marching Method and algorithm based on fluid dynamics. In the current sample we are going to use Fast Marching Method(<code>cv.INPAINT_TELEA</code>): we are consider region of the picture for being inpainted. Algorithm starts from the boundary of this region and goes inside the region gradually filling everything in the boundary first. It takes a small neighbourhood around the pixel on the neigbourhood to be inpainted. This pixel is replaced by normalized weighted sum of all the known pixels in the neigbourhood.
+In this sample we are using next OpenCV functions:
+<code>cv.imread("id of the source image")</code>
+<code>cv.imshow("id of output canvas", name of the showed object)</code>
+<code>cv.line(cv.Mat for drawing, start point as new cv.Point(), end point as new cv.Point(), color as cv.Scalar(), thickness as a  number of pixels)</code>
+<code>cv.cvtColor(source, destination, method)</code>
+<code>cv.inpaint(source image, mask, destination, region of neighborhood as a number of pixels, method)</code>
+</section>
+
+## Instructions
+<section id='instructions'>
+In the current sample we using <code>cv.line()</code> function for damage picture with black lines. You should restore picture with inpaint function.
+Via drawing you can create mask for <code>cv.inpaint()</code> function. Then use "inpaint" function with source picture, created mask, and empty <code>cv.Mat()</code> for saving and showing result.
+</section>
+
+## Tests
+<section id='tests'>
+
+```yml
+tests:
+    - text:  Use <code>cv.ellipse</code> to drow ellips on image
+      testString: assert(code.match(/ellipse/g),'Use <code>cv.ellipse</code> to drow an ellips image');
+    - text:  Use <code>cv.circle</code> to drow ellips on image
+      testString: assert(code.match(/circle/g),'Use <code>cv.circle</code> to drow an ellips image');
+```
+</section>
+
+## Challenge Seed
+
+<section id='challengeSeed'>
+
+<div id='html-seed'>
+
+```html
+
+<canvas id="canvas" >
+</canvas>
+<img id="src" src="http://bit.ly/fcc-relaxing-cat" style="display:none"/>
+<input type="button" id="myButton" value="Test" disabled=true
+onclick="inpaint()"/>
+
+<script type="text/javascript">
+
+var src, mask, lastX, lastY;
+
+window.onload = function()
+{
+context = document.getElementById("canvas").getContext("2d");
+img = document.getElementById("src");
+context.canvas.width = img.width;
+context.canvas.height = img.height;
+context.drawImage(img, 0, 0);
+};
+
+document.getElementById("canvas").onmousedown = function(e)
+{
+  lastX = e.pageX - this.offsetLeft;
+  lastY = e.pageY - this.offsetTop;
+};
+
+document.getElementById("canvas").onmousemove = function(e)
+{
+  if (lastX != undefined)
+  {
+    let x = e.pageX - this.offsetLeft;
+    let y = e.pageY - this.offsetTop;
+    cv.line(src, new cv.Point(lastX, lastY), new cv.Point(x, y),new cv.Scalar(255, 255, 255, 255), 2);
```

Comment:
Looks like we overoptimized the code so we draw on the same Mat which then is used in inpaint.